usersession/userd/launcher.go

```diff
@@ -170,6 +185,187 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface. The desktopFileID is described here:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+func (s *Launcher) OpenDesktopEntryEnv(desktopFileID string, env []string, sender dbus.Sender) *dbus.Error {
```

Comment:
This isn't unit tested. usersession/userd/launcher_test.go should provide some inspiration.

Reply:
@jdstrand I'm having trouble writing a test around this because of the validation it does of the .desktop path. I've got this:

```go
var mircadeDesktop = `[Desktop Entry]
X-SnapInstanceName=mircade
Name=mircade
Exec=env BAMF_DESKTOP_FILE_HINT=/var/lib/snapd/desktop/applications/mircade_mircade.desktop /snap/bin/mircade
Icon=/snap/mircade/143/meta/gui/mircade.png
Comment=Sample confined desktop
Type=Application
Categories=Game
`

func (s *launcherSuite) TestOpenDesktopEntryEnvSucceedsWithGoodDesktopId(c *C) {
	dirs.SetRootDir(c.MkDir())
	cmd := testutil.MockCommand(c, "/snap/bin/mircade", "true")
	defer cmd.Restore()

	deskTopFile := filepath.Join(dirs.SnapDesktopFilesDir, "mircade_mircade.desktop")
	err := os.MkdirAll(filepath.Dir(deskTopFile), 0755)
	c.Assert(err, IsNil)
	err = ioutil.WriteFile(deskTopFile, []byte(strings.Replace(mircadeDesktop, "/var/lib/snapd/desktop/applications", dirs.SnapDesktopFilesDir, -1)), 0644)
	c.Assert(err, IsNil)

	err = s.launcher.OpenDesktopEntryEnv("mircade_mircade.desktop", []string{}, ":some-dbus-sender")
	c.Assert(err, IsNil)
	c.Assert(cmd.Calls(), IsNil)
}
```

Am I missing a simple way to fake the ownership and access of the test directories?

Reply:
You should be able to mock verifyDesktopFileLocation() to return nil or error and test everything else. Then you can mock os.Stat (see MockChownPath() in overlord/configstate/configcore/export_test.go for inspiration).